src/Data/Text/Utf8/AhoCorasick/Automaton.hs

```diff
@@ -374,67 +379,197 @@ uAt = UVector.unsafeIndex
+  consumeInput initialOffset initialRemaining seed initialState
+  where
+    initialState = 0
+
+    Text !u8data !initialOffset !initialRemaining = text
+    AcMachine !values !transitions !offsets !rootAsciiTransitions = machine
 
     -- NOTE: All of the arguments are strict here, because we want to compile
     -- them down to unpacked variables on the stack, or even registers.
-    -- The INLINE / NOINLINE annotations here were added to fix a regression we
-    -- observed when going from GHC 8.2 to GHC 8.6, and this particular
-    -- combination of INLINE and NOINLINE is the fastest one. Removing increases
-    -- the benchmark running time by about 9%.
 
+    -- When we follow an edge, we look in the transition table and do a
+    -- linear scan over all transitions until we find the right one, or
+    -- until we hit the wildcard transition at the end. For 0 or 1 or 2
+    -- transitions that is fine, but the initial state often has more
+    -- transitions, so we have a dedicated lookup table for it, that takes
+    -- up a bit more space, but provides O(1) lookup of the next state. We
+    -- only do this for the first 128 code units (all of ascii).
+
+    -- | Consume a code unit sequence that constitutes a full code point.
+    -- If the code unit at 'offset' is ASCII, we can lower it using 'Utf8.toLowerAscii'.
+    -- Otherwise we have to assume that lowercasing it will change the length of the code unit sequence.
+    -- Therefore we have to invoke 'followCodeUnits', which keeps a queue of code units and looks up their transitions one after the other.
     {-# NOINLINE consumeInput #-}
     consumeInput :: Int -> Int -> a -> State -> a
+    consumeInput _offset 0 acc _state = acc
     consumeInput !offset !remaining !acc !state =
-      let
-        inputCodeUnit = fromIntegral $ indexTextArray u8data offset
-        -- NOTE: Although doing this match here entangles the automaton a bit
-        -- with case sensitivity, doing so is faster than passing in a function
-        -- that transforms each code unit.
-        -- casedCodeUnit = case caseSensitivity of
-        --  IgnoreCase    -> lowerCodeUnit inputCodeUnit
-        --  CaseSensitive -> inputCodeUnit
-      in
-        case remaining of
-          0 -> acc
-          _ -> followEdge (offset + 1) (remaining - 1) acc state inputCodeUnit
+      case caseSensitivity of
+        -- If we are doing case sensitive matching, we can just use the unmodified
+        -- input code units.
+        CaseSensitive ->
+          if cu < asciiCount && state == initialState
+            then lookupRootAsciiTransition (offset + 1) (remaining - 1) acc cu
+            else followCodeUnit (offset + 1) (remaining - 1) acc cu state
+
+        IgnoreCase
+          -- Code point is a single byte ==> It's ASCII, no need to call toLower; we can just use maths :)
+          -- We could check < asciiCount here as well, since there are no UTF-8 code unit sequences starting with 10xxxxxx
```

Comment:
That makes me think, would it be more efficient to just test for the highest bit here (i.e. use `0x80 .&. cu == 0` as condition)?

Reply:
Sadly this doesn't have a noticeable impact on the benchmark (it seems to make it slower but I'd chalk that up to stuff running in the background):

```
./report.py haskell-utf8-lower.stats haskell-utf8-lower-bits.stats 
haskell-utf8-lower.stats:
  mean time: 11.047 ± 0.032 seconds
   min time: 10.704 seconds

haskell-utf8-lower-bits.stats:
  mean time: 11.867 ± 0.050 seconds
   min time: 11.366 seconds
```